File: Data Structures & Algorithms/count-number-of-islands/submission-1.java
class Solution {

    private boolean isValid(int x, int y, int a, int b, char[][] grid){
        if(x<0 || x>=a)
            return false;
        if(y<0 || y>=b)
            return false;

        if(grid[x][y] != '1')
            return false;
        
        return true;
    }

    private void solve(char[][] grid, int x, int y){
        if(!isValid(x, y, grid.length, grid[0].length, grid))
            return;

        grid[x][y]++;

        solve(grid, x+1, y);
        solve(grid, x-1, y);
        solve(grid, x, y+1);
        solve(grid, x, y-1);
    }

    public int numIslands(char[][] grid) {
        int ans = 0;

        for(int i=0; i<grid.length; i++){
            for(int j=0; j<grid[0].length; j++){
                if(grid[i][j] == '1'){
                    ans++;
                    solve(grid, i, j);
                }
            }
        }

        return ans;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-1.cpp
class Solution {
public:

    bool solve(vector<int>& piles, int mid, int h){
        int count = 0;
        for(int it: piles){
            if(it%mid == 0)
                count += it/mid;
            else count += it/mid + 1;
        }

        return count <= h;
    }

    int minEatingSpeed(vector<int>& piles, int h) {
       int l=1, r=*max_element(piles.begin(), piles.end());

        int ans = 0;
       while(l<=r){
        int mid = l + (r-l)/2;

        if(solve(piles, mid, h)){
            ans = mid;
            r=mid-1;
        }
        else
            l=mid+1;
       } 
       return ans;
    }
};
